api/src/main/java/io/grpc/LoadBalancer.java

```diff
@@ -773,6 +775,12 @@ public boolean equals(Object other) {
       Builder() {
       }
 
+      private <T> Builder copyCustomOptions(Object[][] options) {
+        customOptions = new Object[options.length][2];
+        System.arraycopy(customOptions, 0, options, 0, options.length);
```

Comment:
Deep copy?

Reply:
No. Options are intended to be mutable. So keys and values are compared by reference.

Reply:
I'm confused. You used `deepEquals` for comparison which is not compared by identity. And here, coping builder is not for comparison.

Reply:
What does deep copy even mean here? The keys are immutable and the objects are arbitrary. How could we do a deep copy?

Reply:
@dapengzhang0 
In `equals` method, I should do `Arrays.equals` instead of `deepEquals`. Keys and values should be used as identity-based.

Reply:
I can agree using shallow copy for copying builder, but `equals` and `hashCode` should also be shallow for the same reason.

Reply:
>ejona86
What does deep copy even mean here? The keys are immutable and the objects are arbitrary. How could we do a deep copy?

I was just glancing at the code and noticed the inconsistency between copy and equals/hashCode, didn't know about the mutability detail.

Reply:
I don't understand what equals has to do with copy. We can use `equals()` even if we don't deep copy.

Keys should be identity-based. It is unclear if values should. I'd have expected to choose identity if we don't care about equals for custom options and use equals otherwise. This is why I hate defining equals/hashCode...

Reply:
lol. This is very subtle. Strings are always identity-based if we use with literals, while boxed numbers are not. This might bring us some trouble when using this. But I would agree that leaving values identity-based is better.

Reply:
Okay. I'm not hearing anything a reason for using identity for values. Let's use equals().

I'm still +1 for not defining equals though. This doesn't seem like something that anyone would even use during tests. With the `stateListener`, it is highly, highly unlikely to be equal with another instance.